neuralpp/symbolic/expression.py

```diff
@@ -106,13 +106,33 @@ def contains(self, target: Expression) -> bool:
         f(x,f(a,b)).contains(a) == True
         a.contains(a) == True
         """
-        if self == target:
+        if self.syntactic_eq(target):
             return True
-
         for sub_expr in self.subexpressions:
             if sub_expr.contains(target):
                 return True
+        return False
 
+    def structurally_contains(self, target: Expression) -> bool:
```

Comment:
introduced this function because i need this "contains" in normalization of quantifier expression

Reply:
I agree with most of what you said here. The only advantage of syntactic_eq is that it is supposed to be faster. It could be a fast pass in the implementation of structure_eq.

Reply:
In that case the names could be more reflective of that. `syntactic_eq` should be `syntactic_eq_given_same_kind` and `structurally_eq` should be called `syntatically_eq`.